emcc.py

```diff
@@ -2493,7 +2493,7 @@ def generate_html(target, options, js_target, target_basename,
           var memoryInitializer = '%s';
           if (typeof Module['locateFile'] === 'function') {
             memoryInitializer = Module['locateFile'](memoryInitializer);
-          } else if (Module['memoryInitializerPrefixURL']) {
+          } else {
```

Comment:
We don't want to introduce the requirement that these variables would exist unconditionally in `Module`. If they don't, then this would be adding up `undefined` + string.

Reply:
Code in shell.js ensures that either `Module.locateFile` or `Module.memoryInitializerPrefixURL` is present without semantic change, so this condition is simply redundant.

Reply:
The role of the `Module` object has usually been a way for users to add fields to do customization of the loading experience. We have generally had few mandatory fields present there that would cause a build to fail to load. I still think it's better to solve this routing via a mechanism that has less state to it, i.e. just add the single `Module['scriptDirectory']` variable.

Doing this kind of change where we begin to unconditionally depend on `Module['memoryInitializerPrefixURL']` takes the loader code to a direction where it is more dependent on this variable, rather than less, so I think the approach in #5484 is better.

Reply:
@juj, have you seen that `Module['memoryInitializerPrefixURL']` fallbacks to just `Module['scriptDirectory']`?
`Module['scriptDirectory']` on the other hand have solid defaults and you don't need to define it unless you need to change something.

Reply:
Right - I still think it's better to rely less on fields being mandatorily present in `Module`, and rather have an initialization flow that does not require e.g. `Module['memoryInitializerPrefixURL']` always to exist. I know it is initialized unconditionally in the loader, but that does cost code lines, and the initialization flow is asynchronous, so things may get changed in between. It seems more refactor-hardened to me to not require that the older `Module['*PrefixURL']` need to unconditionally exist.

Reply:
Those properties are assigned synchronously on Module initialization before any use, so fortunately asynchronicity is not an issue in this case.

> but that does cost code lines

This can be refactored later, most of the conditional statements will go away when `Module['*PrefixURL']` is dropped somewhere in future. It will boil down to single Closure-compressable property for everything.